### PR TITLE
Revcontent AdNetwork Service

### PR DESCRIPTION
**Pull request by e2corporation**
Integrating **Revcontent** as a Native Ad provider type for AMPHTML.

### CODE REVIEW

**Comment by e2corporation** [comment]
A Live build of this branch (with some environment overrides in place) may be viewed at:

**Revcontent + AMP** (SSL)
https://amphtml.bluestallion.net/examples/revcontent.amp.html

Please direct any build issues with this branch to [ julien@revcontent.com](mailto: julien@revcontent.com)

File: examples/revcontent.amp.html [hunk 1]
```diff
@@ -0,0 +1,120 @@
+<!doctype html>
```
Comment (cramforce) [inline examples/revcontent.amp.html]: Please add to `ads.amp.html` instead.

File: src/3p.js [hunk 2]
```diff
@@ -114,7 +114,7 @@ export function validateSrcPrefix(prefix, src) {
     prefix = [prefix];
   }
 
-  for (const p of prefix) {
+  for (const p in prefix) {
```
Comment (cramforce) [inline src/3p.js]: Please use a counting for-loop instead.

File: ads/revcontent.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+import {writeScript, checkData} from '../src/3p';
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function revcontent(global, data) {
+  checkData(data, ['id', 'width', 'height', 'endpoint', 'ssl', 'wrapper']);
+  (function(window, document, data) {
```
Comment (cramforce) [inline ads/revcontent.js]: Please manually inline this function. It is confusing.

File: ads/revcontent.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    const revampServiceUrl = 'https://labs-cdn.revcontent.com/build/amphtml/revcontent.amp.min.js';
```
Comment (cramforce) [inline ads/revcontent.js]: Please add this URL to `_config.js`

File: ads/revcontent.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        data-endpoint="trends.revcontent.com"
+        data-ssl="true"
+        data-id="203">
+
+</amp-ad>
+```
+
+## Configuration
+
+For semantics of configuration, please see ad network documentation.
```
Comment (cramforce) [inline ads/revcontent.md]: Please ad a link

File: ads/revcontent.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+Supported parameters:
+
+- wrapper. Must be a valid dom.div id string
```
Comment (cramforce) [inline ads/revcontent.md]: Please add names in `

File: src/mode.js [hunk 7]
```diff
@@ -51,18 +51,19 @@ export function setModeForTesting(m) {
  * @return {!ModeDef}
  */
 function getMode_() {
-  const isLocalDev = (location.hostname == 'localhost' ||
-      (location.ancestorOrigins && location.ancestorOrigins[0] &&
-          location.ancestorOrigins[0].indexOf('http://localhost:') == 0)) &&
+  const hostname = location.hostname;
```
Comment (cramforce) [inline src/mode.js]: Please revert this change.

File: src/3p.js (hunk 2 above)
Comment (e2corporation) [inline src/3p.js]: There are additional issues to address with this method, please see Issue #2007 for more information. I will attempt resolve it with this branch unless your team is already aware of the issue and there is another fix under development.

File: ads/revcontent.md (hunk 6 above)
Comment (e2corporation) [inline ads/revcontent.md]: Our *readme.md* file was based off of another template from several revisions ago, it will update to match layout of the other readme files.

File: src/mode.js (hunk 7 above)
Comment (e2corporation) [inline src/mode.js]: Will be reverted, this change was to allow flexible local development, the codebase has assumptions that host "**ads.localhost**" will always be used for development which makes staging/testing the stack cumbersome.

Are there gulp environment flags available to enforce a local development environment?

Comment (cramforce) [inline src/mode.js]: Could you file an issue for correctly picking up dev mode in 3p iframes? Thanks!

File: ads/revcontent.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+- endpoint. Must be a valid Trends Server *required*
+- id. Must be a valid Widget ID *required* 
+- ssl. Must be set to "true" for SSL mode *required*
```
Comment (cramforce) [inline ads/revcontent.md]: Maybe hardcode this in JS and make it not configurable.

File: ads/revcontent.md (hunk 6 above)
Comment (cramforce) [inline ads/revcontent.md]: It happens that we miss stuff like that. Just looks much better :)

File: src/mode.js (hunk 7 above)
Comment (e2corporation) [inline src/mode.js]: Sure, Issue #2008 Filed to request this feature. https://github.com/ampproject/amphtml/issues/2008

File: ads/revcontent.md (hunk 8 above)
Comment (e2corporation) [inline ads/revcontent.md]: Would it be ok if I remove **`ssl`** as user-facing option from our readme? **SSL** will still default to true if not provided (Pending a deploy on our CDN for our service file), so this parameter would be used to turn SSL off in most cases.

Comment (cramforce) [inline ads/revcontent.md]: Sounds good.

File: ads/_config.js [hunk 9]
```diff
@@ -44,19 +45,25 @@ export const adPreconnect = {
   taboola: [
     'https://cdn.taboola.com',
     'https://trc.taboola.com',
-    'https://images.taboola.com',
```
Comment (cramforce) [inline ads/_config.js]: We use trailing kommas. Will be enforced soon. Please restore.

Comment (e2corporation) [inline ads/_config.js]: Trailing commas restored.

File: examples/revcontent.amp.html (hunk 1 above)
Comment (e2corporation) [inline examples/revcontent.amp.html]: Example code migrated to main ads.amp.html file.

File: ads/revcontent.js (hunk 3 above)
Comment (e2corporation) [inline ads/revcontent.js]: This has been cleaned up, extra closure removed.

File: ads/revcontent.md (hunk 6 above)
Comment (e2corporation) [inline ads/revcontent.md]: README updated to match current template.

File: ads/revcontent.md (hunk 8 above)
Comment (e2corporation) [inline ads/revcontent.md]: **SSL** option stripped from example markup.

File: ads/_config.js [hunk 10]
```diff
@@ -57,6 +58,12 @@ export const adPreconnect = {
     'https://s.yieldmo.com',
     'https://ads.yieldmo.com',
   ],
+  revcontent: [
+    'https://trends.revcontent.com',
+    'https://cdn.revcontent.com',
+    'https://img.revcontent.com',
+    'https://labs-cdn.revcontent.com',
```
Comment (cramforce) [inline ads/_config.js]: You can remove this. We always preconnect to hosts that we prefetch from.

File: ads/revcontent.md (hunk 5 above)
Comment (cramforce) [inline ads/revcontent.md]: Friendly ping

File: src/3p.js [hunk 11]
```diff
@@ -109,18 +109,25 @@ function executeAfterWriteScript(win, fn) {
...
-
+  const validPrefixes = [];
   if (!isArray(prefix)) {
     prefix = [prefix];
   }
-
-  for (const p of prefix) {
-    if (src.indexOf(p) === 0) {
+  const totalPrefixes = prefix.length;
+  if (src !== undefined) {
```
Comment (cramforce) [inline src/3p.js]: src cannot be undefined according to the type. We should definitely fail in that case.

File: src/3p.js [hunk 12]
```diff
@@ -109,18 +109,25 @@ function executeAfterWriteScript(win, fn) {
...
+    let p = 0;
```
Comment (cramforce) [inline src/3p.js]: inline into for-statement

File: src/3p.js [hunk 13]
```diff
@@ -109,18 +109,25 @@ function executeAfterWriteScript(win, fn) {
...
+    for (p = 0; p <= totalPrefixes; p++) {
+      const protocolIndex = src.toLowerCase().indexOf(prefix[p]);
```
Comment (cramforce) [inline src/3p.js]: Remove `toLowerCase`

File: src/3p.js [hunk 14]
```diff
@@ -109,18 +109,25 @@ function executeAfterWriteScript(win, fn) {
...
 export function validateSrcPrefix(prefix, src) {
-
+  const validPrefixes = [];
   if (!isArray(prefix)) {
     prefix = [prefix];
   }
-
-  for (const p of prefix) {
-    if (src.indexOf(p) === 0) {
+  const totalPrefixes = prefix.length;
```
Comment (cramforce) [inline src/3p.js]: inline var into single usage

File: src/3p.js [hunk 15]
```diff
@@ -109,18 +109,25 @@ function executeAfterWriteScript(win, fn) {
...
+  if (src !== undefined) {
+    let p = 0;
+    for (p = 0; p <= totalPrefixes; p++) {
+      const protocolIndex = src.toLowerCase().indexOf(prefix[p]);
+      if (protocolIndex == 0) {
+        validPrefixes.push(p.toString());
```
Comment (cramforce) [inline src/3p.js]: Why not immediately return here?

Comment (e2corporation) [inline src/3p.js]: Since this function was design to validate 1 or more prefixes ['https:','http:'], in the event we do have matches we want to keep track of it separately, especially if an invalid protocol is presented first, i.e. **http** in the list of prefixes. 

So the flow is structured this way to go after one of the original problems with this function (#2007), that failed it's own test case when accepting 2 or prefixes, causing it to throw errors during the build even when 1 valid prefix was in the list.

File: src/3p.js (hunk 13 above)
Comment (e2corporation) [inline src/3p.js]: Even though indexOf is case-sensitive? Is **src** always going to be provided in lowercase?

File: ads/revcontent.md (hunk 5 above)
Comment (e2corporation) [inline ads/revcontent.md]: Good to go.

File: ads/_config.js (hunk 10 above)
Comment (e2corporation) [inline ads/_config.js]: So we just need the one entry above in **adPrefech**? Those host mentioned in pre-fetch is only used for the js service file.

File: src/3p.js (hunk 12 above)
Comment (e2corporation) [inline src/3p.js]: if we inline P we get an error on the Build about p being read-only, See Build No. 3858 https://travis-ci.org/ampproject/amphtml/builds/109090168

screen shot 2016-02-14 at 12 20 11 pm


I can still inline this if you would like to see if the build passes.

**Comment by e2corporation** [comment]
@googlebot CLA Signed on behalf of Corporation "Revcontent, LLC"

File: src/3p.js (hunk 13 above)
Comment (cramforce) [inline src/3p.js]: This is supposed to be case sensitive. If anything only the origin should be case-insensitive. The path must be case sensitive.

File: src/3p.js (hunk 15 above)
Comment (cramforce) [inline src/3p.js]: The intent here is "one of" semantics. It is sufficient that one prefix matches.

File: src/3p.js (hunk 13 above)
Comment (e2corporation) [inline src/3p.js]: This action is only lower-casing to validate against the **protocol**/prefix not the entire path, A User agent should not treat HTTPS:// and https:// differently. Ideally both the prefix[p] and the source should be lower-cased then validated for protocol.

Removing on next commit.

File: src/3p.js (hunk 11 above)
Comment (e2corporation) [inline src/3p.js]: Updated with previous commits to throw an error after the if check.

File: src/3p.js (hunk 15 above)
Comment (e2corporation) [inline src/3p.js]: validateSrcPrefix has been cleaned up with Commit 8d7a5a8

File: ads/_config.js (hunk 10 above)
Comment (e2corporation) [inline ads/_config.js]: Is **_config.js** good to go as-is?

Comment (cramforce) [inline ads/_config.js]: You can remove the `'https://labs-cdn.revcontent.com'` entry here, as all entries in the prefetch list also always get a get a preconnect.

File: examples/ads.amp.html [hunk 16]
```diff
@@ -176,5 +183,18 @@
       type="yieldmo"
       data-ymid="1349317029731662884">
   </amp-ad>
+
+  <h2>Revconntent Widget with placeholder and fallback</h2>
```
Comment (cramforce) [inline examples/ads.amp.html]: Typo?

File: ads/_config.js (hunk 10 above)
Comment (e2corporation) [inline ads/_config.js]: Updated.

File: examples/ads.amp.html (hunk 16 above)
Comment (e2corporation) [inline examples/ads.amp.html]: Thanks for the catch, typo fixed.

File: src/3p.js [hunk 17]
```diff
@@ -109,17 +109,20 @@ function executeAfterWriteScript(win, fn) {
...
     prefix = [prefix];
   }
-
-  for (const p of prefix) {
-    if (src.indexOf(p) === 0) {
-      return;
+  const totalPrefixes = prefix.length;
+  if (src !== undefined) {
+    let p = 0;
+    for (p = 0; p <= totalPrefixes; p++) {
```
Comment (cramforce) [inline src/3p.js]: Remove previous line and make this `for (var p = 0; …)`

File: src/3p.js [hunk 18]
```diff
@@ -109,17 +109,20 @@ function executeAfterWriteScript(win, fn) {
...
 export function validateSrcPrefix(prefix, src) {
-
   if (!isArray(prefix)) {
     prefix = [prefix];
   }
-
-  for (const p of prefix) {
-    if (src.indexOf(p) === 0) {
-      return;
+  const totalPrefixes = prefix.length;
```
Comment (cramforce) [inline src/3p.js]: Inline

File: src/3p.js [hunk 19]
```diff
@@ -109,17 +109,20 @@ function executeAfterWriteScript(win, fn) {
...
+  if (src !== undefined) {
+    let p = 0;
+    for (p = 0; p <= totalPrefixes; p++) {
+      const protocolIndex = src.indexOf(prefix[p]);
+      if (protocolIndex == 0) {
+        return;
+      }
     }
+    throw new Error('Invalid src ' + src);
```
Comment (cramforce) [inline src/3p.js]: You can delete this line

File: src/3p.js [hunk 20]
```diff
@@ -109,17 +109,20 @@ function executeAfterWriteScript(win, fn) {
...
   }
-
-  for (const p of prefix) {
-    if (src.indexOf(p) === 0) {
-      return;
+  const totalPrefixes = prefix.length;
+  if (src !== undefined) {
+    let p = 0;
+    for (p = 0; p <= totalPrefixes; p++) {
+      const protocolIndex = src.indexOf(prefix[p]);
```
Comment (cramforce) [inline src/3p.js]: Inline

**Comment by cramforce** [comment]
Some comments, but otherwise this is ready to go. Please squash your commits: https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/

File: src/3p.js (hunk 19 above)
Comment (e2corporation) [inline src/3p.js]: This was to handle the previous request to throw an error if src is not defined..

File: src/3p.js (hunk 17 above)
Comment (e2corporation) [inline src/3p.js]: Travis CI will error about usage of **var**, and instead will request **const** or **let**

File: src/3p.js (hunk 19 above)
Comment (cramforce) [inline src/3p.js]: It falls through to the next line

File: src/3p.js (hunk 17 above)
Comment (cramforce) [inline src/3p.js]: Sorry, `for (let p = 0; …)`

File: src/3p.js (hunk 19 above)
Comment (e2corporation) [inline src/3p.js]: the first throw is if src is defined and no matches are found. The last throw is if src is undefined, try the sample below in the console which has the first throw removed, you'll see that the error is not throw when http does not exist in https://example.org

```
var toString_ = Object.prototype.toString;

function toString(value) {
  return toString_.call(value);
}

function isArray(value) {
  return toString(value) === '[object Array]';
}

function validateSrcPrefix(prefix, src) {
  if (!isArray(prefix)) {
    prefix = [prefix];
  }
  const totalPrefixes = prefix.length;
  if (src !== undefined) {
    for (var p = 0; p <= totalPrefixes; p++) {
      const protocolIndex = src.indexOf(prefix[p]);
      if (protocolIndex == 0) {
        return;
      }
    }
  }
  throw new Error('Invalid src ' + src);
}
```

Comment (cramforce) [inline src/3p.js]: Not sure what you are saying. That code looks good (modulo my other style comments.) and works. Feel free to add a failing test case to prove otherwise.

Comment (e2corporation) [inline src/3p.js]: I forgot to actually run the validate call, snippet update above. Does the above sample raise an error on the console? It will output undefined and not throw an error.

Comment (cramforce) [inline src/3p.js]: Sure it throws, all code paths that do not return explicitly throw.

File: src/3p.js (hunk 17 above)
Comment (e2corporation) [inline src/3p.js]: Updated.

File: src/3p.js (hunk 19 above)
Comment (e2corporation) [inline src/3p.js]: Updated.

File: src/3p.js (hunk 18 above)
Comment (e2corporation) [inline src/3p.js]: Updated.

**Comment by e2corporation** [comment]
@cramforce Thanks Malte, in regards to the squash, my develop tree has had several commits and a few merges in from master. All of our commits are not in direct sequence, and doing some initial tests we'll have manual merge conflicts in **integration.js **and **3p.js **.

I was thinking I'd revert my develop back to 1 before my first ever commit in the tree, rebase from amphtml/master then apply our stable **develop-1.0** as a Squashed commit.

Could a new pull request would be a fallback option as well?

**Comment by e2corporation** [comment]
@cramforce A new Release PR has been created in the mean time, rebasing onto this branch is causing several manual conflicts. See Pull Request No. 2022

**PR 2022**
https://github.com/ampproject/amphtml/pull/2022

**Comment by cramforce** [comment]
Closing in favor of #2022
